src/main/jwks.ts

```diff
@@ -0,0 +1,126 @@
+import { Request } from '@automationcloud/request';
+import assert from 'assert';
+import { Exception } from './exception';
+import { Response } from 'node-fetch';
+
+export class JwksClient {
+    protected _cache: Map<string, JwksCache> = new Map();
+
+    constructor(protected options: JwksOptions) {}
+
+    protected get request() {
+        return new Request({
+            baseUrl: this.options.url,
+            retryAttempts: this.options.retryAttempts ?? 3,
+        });
+    }
+
+    async getSigningKey(): Promise<string> {
+        const alg = this.options.algorithm;
+        const cached = this.cache.get(alg);
+        if (cached) {
+            return cached;
+        }
+
+        const res = await this.request.get('');
+        const keys = this.decodeKeys(res);
+        const key = keys.find(k => k.alg === this.options.algorithm);
+        if (!key) {
+            throw new Exception({
+                name: 'SigningKeyNotFoundError',
+                message: 'Expected signing key not found from response',
+            });
+        }
+
+        this.cache.set(key.alg, key.k);
+
+        return key.k;
+    }
+
+    get cache() {
+        return {
+            get: (key: string) => {
+                const v = this._cache.get(key);
+                if (!v || v.expiresAt < Date.now()) {
+                    this._cache.delete(key);
+                    return null;
+                }
+
+                return v.value;
+            },
+
+            set: (key: string, value: string, maxAge?: number) => {
+                const ttl = maxAge ??
+                    this.options.cacheMaxAge ??
+                    60 * 1000;
+
+                const expiresAt = Date.now() + ttl;
+                this._cache.set(key, { expiresAt, value });
+            },
+
+            clear: () => {
+                this._cache = new Map();
+            }
+        }
+    }
+
+    protected decodeKeys(res: { [k: string]: any }): AcSigningKey[] {
+        const { keys } = res;
+        assert(res.keys && Array.isArray(keys), '\'keys\' expected to be array');
```

Comment:
I suggest we use `ajv` for validating responses (nitpick: decoding is _technically_ an incorrect term, because no data transformation/conversion takes place :P). I think JSON schema (despite some of its shortcomings) provides a much better way to validate, mainly because custom predicates are a bit anecdotal and harder to read/consume.

On a related: originally `assert` wasn't supposed to be a runtime check, rather it was a development-time construct which served to document and enforce invariants in design-by-contract — kind of like unit tests embedded in the code and stripped by compiler. For example, in Java assertions would only execute with a special flag, and regular `catch (Exception e)` wouldn't catch assertion errors, because they are not subclasses of `Exception` (which was/is convenient to spot bugs). That's a bit of a CS history lesson 👨‍🏫  — so no action is required, but kinda helps with my point above 😁